docs/the-new-architecture/custom-cxx-types.md

```diff
@@ -0,0 +1,417 @@
+
+It is also possible to have functions that return custom types.
+
+### 2. Define the bridging code
+
+From the Address type defined in the specs, Codegen will generate two helper types: `NativeSampleModuleAddress` and `NativeSampleModuleAddressBridging`.
+
+The first type is the definition of the `Address`. The second type contains all the plumbing to bridge the custom type from JS to C++ and viceversa. The only extra step we need to add is to define the `Bridging` structure that extends the `NativeSampleModuleAddressBridging` type.
+
+1. Open the `shared/NativeSampleModule.h` file
+2. Add the following code in the file:
+
+```diff title="NativeSampleModule.h (Bridging the Address type)"
+#include "Int64.h"
+#include <memory>
+#include <string>
+
+namespace facebook::react {
++  using Address = NativeSampleModuleAddress<std::string, int32_t, bool>;
+
++  template <>
++  struct Bridging<Address>
++      : NativeSampleModuleAddressBridging<Address> {};
+  // ...
+}
+```
+
+This code defines an `Address` typealias for the generic type `NativeSampleModuleAddress`. **The order of the generics matters**: the first template argument refers to the first data type of the struct, the second refers to the second, and so forth.
+
+Then, the code adds the `Bridging` specialization for the new `Address` type, by extending `NativeSampleModuleAddressBridging` that is generated by Codegen.
+
+:::note
+There is a convention that is followed to generate this types:
+
+- The first part of the name is always the type of the module. `NativeSampleModule`, in this example.
+- The second part of the name is always the name of the JS type defined in the specs. `Address`, in this example.
+  :::
+
+### 3. Implement the Native Code
+
+Now, we need to implement the `validateAddress` function in C++. First, we need to add the function declaration into the `.h` file, and then we can implement it in the `.cpp` file.
+
+1. Open the `shared/NativeSampleModule.h` file and add the function definition
+
+```diff title="NativeSampleModule.h (validateAddress function prototype)"
+  std::string reverseString(jsi::Runtime& rt, std::string input);
+
++  bool validateAddress(jsi::Runtime &rt, jsi::Object input);
+};
+
+} // namespace facebook::react
+```
+
+2. Open the `shared/NativeSampleModule.cpp` file and add the function implementation
+
+```c++ title="NativeSampleModule.cpp (validateAddress implementation)"
+bool NativeSampleModule::validateAddress(jsi::Runtime &rt, jsi::Object input) {
+  std::string street = input.getProperty(rt, "street").asString(rt).utf8(rt);
+  int32_t number = input.getProperty(rt, "num").asNumber();
+
+  return !street.empty() && number > 0;
+}
+```
+
+In the implementation, the object that represents the `Address` is a `jsi::Object`. To extract the values from this object, we need to use the accessors provided by `JSI` (If you want to learn more about `JSI`, have a look at this [great talk](https://youtu.be/oLmGInjKU2U?feature=shared) from App.JS 2024):
```

Comment:
Not sure I would link the talk here. Maybe in a section "Learn more" at the bottom?

Reply:
This is the only point in the whole document where we talk about JSI. We can definitely add a `Learn More` at the bottom of the page, but it feels decontextualized to me...

What if we add it right below this explanation? PErhaps in a `note` box

Reply:
Yeah the note box also works